File: web/src/stores/lifecycleStatus.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { lifecycleStatusAPI } from '@/services/api'

export interface LifecycleStatus {
  id: string
  name: string
  display_name: string
  description?: string
  color?: string
  icon?: string
  sort_order: number
  is_active: boolean
  is_system: boolean
  created_at: string
  updated_at: string
  created_by: string
}

export interface LifecycleStatusListResponse {
  lifecycle_statuses: LifecycleStatus[]
  page: number
  limit: number
  total: number
  total_pages: number
}

export interface LifecycleStatusUsage {
  lifecycle_status: LifecycleStatus
  usage_count: number
}

export interface LifecycleStatusUsageResponse {
  total_cis: number
  cis_with_status: number
  cis_without_status: number
  status_usage: LifecycleStatusUsage[]
  status_distribution: Record<string, number>
}

export interface CIStatusDistribution {
  status_name: string
  display_name: string
  color: string
  icon: string
  count: number
  percentage: number
}

export interface CreateLifecycleStatusRequest {
  name: string
  display_name: string
  description?: string
  color?: string
  icon?: string
  sort_order?: number
}

export interface UpdateLifecycleStatusRequest {
  display_name?: string
  description?: string
  color?: string
  icon?: string
  sort_order?: number
  is_active?: boolean
}

export const useLifecycleStatusStore = defineStore('lifecycleStatus', () => {
  // State
  const lifecycleStatuses = ref<LifecycleStatus[]>([])
  const activeLifecycleStatuses = ref<LifecycleStatus[]>([])
  const currentLifecycleStatus = ref<LifecycleStatus | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  // Actions
  const listLifecycleStatuses = async (
    page: number = 1,
    limit: number = 20,
    filters?: {
      search?: string
      is_active?: boolean
      is_system?: boolean
      sort?: string
      order?: string
    }
  ): Promise<LifecycleStatusListResponse> => {
    try {
      loading.value = true
      error.value = null

      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString()
      })

      if (filters) {
        if (filters.search) {
          params.append('search', filters.search)
        }
        if (filters.is_active !== undefined) {
          params.append('is_active', filters.is_active.toString())
        }
        if (filters.is_system !== undefined) {
          params.append('is_system', filters.is_system.toString())
        }
        if (filters.sort) {
          params.append('sort', filters.sort)
        }
        if (filters.order) {
          params.append('order', filters.order)
        }
      }

      const response = await lifecycleStatusAPI.list(params)
      return response.data
    } catch (err: any) {
      error.value = err.message || 'Failed to load lifecycle statuses'
      throw err
    } finally {
      loading.value = false
    }
  }

  const getLifecycleStatus = async (id: string): Promise<LifecycleStatus> => {
    try {
      loading.value = true
      error.value = null

      const response = await lifecycleStatusAPI.get(id)
      currentLifecycleStatus.value = response.data
      return response.data
    } catch (err: any) {
      error.value = err.message || 'Failed to load lifecycle status'
      throw err
    } finally {
      loading.value = false
    }
  }

  const getActiveLifecycleStatuses = async (): Promise<LifecycleStatus[]> => {
    try {
      loading.value = true
      error.value = null

      const response = await lifecycleStatusAPI.getActive()
      activeLifecycleStatuses.value = response.data
      return response.data
    } catch (err: any) {
      error.value = err.message || 'Failed to load active lifecycle statuses'
      throw err
    } finally {
      loading.value = false
    }
  }

  const createLifecycleStatus = async (data: CreateLifecycleStatusRequest): Promise<LifecycleStatus> => {
    try {
      loading.value = true
      error.value = null

      console.log('Creating Lifecycle Status with data:', data)

      const response = await lifecycleStatusAPI.create(data)
      const newLifecycleStatus = response.data

      console.log('Lifecycle Status created successfully:', newLifecycleStatus)

      // Add to local state
      lifecycleStatuses.value.push(newLifecycleStatus)

      // Update active statuses if it's active
      if (newLifecycleStatus.is_active) {
        activeLifecycleStatuses.value.push(newLifecycleStatus)
        // Sort by sort_order
        activeLifecycleStatuses.value.sort((a, b) => a.sort_order - b.sort_order)
      }

      return newLifecycleStatus
    } catch (err: any) {
      console.error('Failed to create Lifecycle Status:', err)

      // Enhanced error handling
      if (err.response?.status === 422) {
        const errorDetails = err.response.data?.error?.details
        if (errorDetails) {
          const errorMessages = Object.values(errorDetails).flat()
          error.value = `Validation errors: ${errorMessages.join('. ')}`
        } else {
          error.value = 'Validation failed. Please check your input.'
        }
      } else if (err.response?.status === 409) {
        error.value = 'A lifecycle status with this name already exists'
      } else if (err.response?.status >= 500) {
        const serverMessage = err.response.data?.error?.message || 'Server error occurred'
        error.value = `${serverMessage}. Please try again later.`
      } else {
        error.value = err.message || 'Failed to create lifecycle status'
      }

      throw err
    } finally {
      loading.value = false
    }
  }

  const updateLifecycleStatus = async (id: string, data: UpdateLifecycleStatusRequest): Promise<LifecycleStatus> => {
    try {
      loading.value = true
      error.value = null

      console.log('Updating Lifecycle Status:', { id, data })

      // Filter out undefined values to avoid sending unnecessary data
      const updateData: any = {}
      if (data.display_name !== undefined) {
        updateData.display_name = data.display_name
      }
      if (data.description !== undefined) {
        updateData.description = data.description
      }
      if (data.color !== undefined) {
        updateData.color = data.color
      }
      if (data.icon !== undefined) {
        updateData.icon = data.icon
      }
      if (data.sort_order !== undefined) {
        updateData.sort_order = data.sort_order
      }
      if (data.is_active !== undefined) {
        updateData.is_active = data.is_active
      }

      console.log('Sending update request:', { id, data: updateData })

      const response = await lifecycleStatusAPI.update(id, updateData)
      const updatedLifecycleStatus = response.data

      console.log('Lifecycle Status updated successfully:', updatedLifecycleStatus)

      // Update local state
      const index = lifecycleStatuses.value.findIndex(ls => ls.id === id)
      if (index !== -1) {
        lifecycleStatuses.value[index] = updatedLifecycleStatus
      }

      // Update active statuses
      const activeIndex = activeLifecycleStatuses.value.findIndex(ls => ls.id === id)
      if (activeIndex !== -1) {
        if (updatedLifecycleStatus.is_active) {
          activeLifecycleStatuses.value[activeIndex] = updatedLifecycleStatus
          // Sort by sort_order
          activeLifecycleStatuses.value.sort((a, b) => a.sort_order - b.sort_order)
        } else {
          // Remove from active if deactivated
          activeLifecycleStatuses.value.splice(activeIndex, 1)
        }
      } else if (updatedLifecycleStatus.is_active) {
        // Add to active if newly activated
        activeLifecycleStatuses.value.push(updatedLifecycleStatus)
        activeLifecycleStatuses.value.sort((a, b) => a.sort_order - b.sort_order)
      }

      if (currentLifecycleStatus.value?.id === id) {
        currentLifecycleStatus.value = updatedLifecycleStatus
      }

      return updatedLifecycleStatus
    } catch (err: any) {
      console.error('Failed to update Lifecycle Status:', err)

      // Enhanced error handling
      if (err.response?.status === 422) {
        const errorDetails = err.response.data?.error?.details
        if (errorDetails) {
          const errorMessages = Object.values(errorDetails).flat()
          error.value = `Validation errors: ${errorMessages.join('. ')}`
        } else {
          error.value = 'Validation failed. Please check your input.'
        }
      } else if (err.response?.status === 404) {
        error.value = 'Lifecycle status not found'
      } else if (err.response?.status === 403) {
        error.value = 'Cannot modify system lifecycle statuses'
      } else if (err.response?.status >= 500) {
        const serverMessage = err.response.data?.error?.message || 'Server error occurred'
        error.value = `${serverMessage}. Please try again later.`
      } else {
        error.value = err.message || 'Failed to update lifecycle status'
      }

      throw err
    } finally {
      loading.value = false
    }
  }

  const deleteLifecycleStatus = async (id: string): Promise<void> => {
    try {
      loading.value = true
      error.value = null

      await lifecycleStatusAPI.delete(id)

      // Remove from local state
      lifecycleStatuses.value = lifecycleStatuses.value.filter(ls => ls.id !== id)
      activeLifecycleStatuses.value = activeLifecycleStatuses.value.filter(ls => ls.id !== id)

      if (currentLifecycleStatus.value?.id === id) {
        currentLifecycleStatus.value = null
      }
    } catch (err: any) {
      console.error('Failed to delete Lifecycle Status:', err)

      // Enhanced error handling
      if (err.response?.status === 404) {
        error.value = 'Lifecycle status not found'
      } else if (err.response?.status === 403) {
        error.value = 'Cannot delete system lifecycle statuses'
      } else if (err.response?.status === 409) {
        error.value = 'Cannot delete lifecycle status that is in use'
      } else if (err.response?.status >= 500) {
        const serverMessage = err.response.data?.error?.message || 'Server error occurred'
        error.value = `${serverMessage}. Please try again later.`
      } else {
        error.value = err.message || 'Failed to delete lifecycle status'
      }

      throw err
    } finally {
      loading.value = false
    }
  }

  const getLifecycleStatusUsage = async (): Promise<LifecycleStatusUsageResponse> => {
    try {
      const response = await lifecycleStatusAPI.getUsage()
      return response.data
    } catch (err: any) {
      console.error('Failed to load lifecycle status usage statistics:', err)
      error.value = err.message || 'Failed to load lifecycle status usage statistics'
      throw err
    }
  }

  const getCIStatusDistribution = async (): Promise<CIStatusDistribution[]> => {
    try {
      const response = await lifecycleStatusAPI.getDistribution()
      return response.data
    } catch (err: any) {
      console.error('Failed to load CI status distribution:', err)
      error.value = err.message || 'Failed to load CI status distribution'
      throw err
    }
  }

  const clearError = () => {
    error.value = null
  }

  const clearCurrentLifecycleStatus = () => {
    currentLifecycleStatus.value = null
  }

  return {
    // State
    lifecycleStatuses,
    activeLifecycleStatuses,
    currentLifecycleStatus,
    loading,
    error,

    // Actions
    listLifecycleStatuses,
    getLifecycleStatus,
    getActiveLifecycleStatuses,
    createLifecycleStatus,
    updateLifecycleStatus,
    deleteLifecycleStatus,
    getLifecycleStatusUsage,
    getCIStatusDistribution,
    clearError,
    clearCurrentLifecycleStatus
  }
})

File: internal/ci/lifecycle_status.go
package ci

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

// LifecycleStatus represents a lifecycle status for CIs
type LifecycleStatus struct {
	ID          uuid.UUID  `json:"id" db:"id"`
	Name        string     `json:"name" db:"name"`
	DisplayName string     `json:"display_name" db:"display_name"`
	Description *string    `json:"description,omitempty" db:"description"`
	Color       *string    `json:"color,omitempty" db:"color"`
	Icon        *string    `json:"icon,omitempty" db:"icon"`
	SortOrder   int        `json:"sort_order" db:"sort_order"`
	IsActive    bool       `json:"is_active" db:"is_active"`
	IsSystem    bool       `json:"is_system" db:"is_system"`
	CreatedAt   time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt   *time.Time `json:"updated_at,omitempty" db:"updated_at"`
	CreatedBy   uuid.UUID  `json:"created_by" db:"created_by"`
	UpdatedBy   *uuid.UUID `json:"updated_by,omitempty" db:"updated_by"`
}

// Request/Response DTOs for Lifecycle Statuses

// CreateLifecycleStatusRequest represents a request to create a new lifecycle status
type CreateLifecycleStatusRequest struct {
	Name        string  `json:"name" binding:"required,min=2,max=100"`
	DisplayName string  `json:"display_name" binding:"required,min=2,max=100"`
	Description *string `json:"description,omitempty" binding:"omitempty,max=500"`
	Color       *string `json:"color,omitempty" binding:"omitempty,len=7"`
	Icon        *string `json:"icon,omitempty" binding:"omitempty,max=50"`
	SortOrder   *int    `json:"sort_order,omitempty" binding:"omitempty,min=0"`
}

// UpdateLifecycleStatusRequest represents a request to update an existing lifecycle status
type UpdateLifecycleStatusRequest struct {
	DisplayName *string `json:"display_name,omitempty" binding:"omitempty,min=2,max=100"`
	Description *string `json:"description,omitempty" binding:"omitempty,max=500"`
	Color       *string `json:"color,omitempty" binding:"omitempty,len=7"`
	Icon        *string `json:"icon,omitempty" binding:"omitempty,max=50"`
	SortOrder   *int    `json:"sort_order,omitempty" binding:"omitempty,min=0"`
	IsActive    *bool   `json:"is_active,omitempty"`
}

// LifecycleStatusListResponse represents a paginated list of lifecycle statuses
type LifecycleStatusListResponse struct {
	LifecycleStatuses []LifecycleStatus `json:"lifecycle_statuses"`
	Page              int               `json:"page"`
	Limit             int               `json:"limit"`
	Total             int64             `json:"total"`
	TotalPages        int               `json:"total_pages"`
}

// ListLifecycleStatusFilters represents filtering options for listing lifecycle statuses
type ListLifecycleStatusFilters struct {
	Search   string  `json:"search,omitempty"`
	IsActive *bool   `json:"is_active,omitempty"`
	IsSystem *bool   `json:"is_system,omitempty"`
	Sort     string  `json:"sort,omitempty"`
	Order    string  `json:"order,omitempty"`
}

// LifecycleStatusUsage represents usage statistics for a single lifecycle status
type LifecycleStatusUsage struct {
	LifecycleStatus LifecycleStatus `json:"lifecycle_status"`
	UsageCount     int64            `json:"usage_count"`
}

// LifecycleStatusUsageResponse represents usage statistics for all lifecycle statuses
type LifecycleStatusUsageResponse struct {
	TotalCIs          int64                    `json:"total_cis"`
	CIsWithStatus     int64                    `json:"cis_with_status"`
	CIsWithoutStatus  int64                    `json:"cis_without_status"`
	StatusUsage       []LifecycleStatusUsage   `json:"status_usage"`
	StatusDistribution map[string]int64        `json:"status_distribution"`
}

// CIStatusDistribution represents CI status distribution for dashboard
type CIStatusDistribution struct {
	StatusName     string `json:"status_name"`
	DisplayName    string `json:"display_name"`
	Color          string `json:"color"`
	Icon           string `json:"icon"`
	Count          int64  `json:"count"`
	Percentage     float64 `json:"percentage"`
}

// CIStatusDistributionByType represents CI status distribution grouped by CI type
type CIStatusDistributionByType struct {
	CITypeName string                `json:"ci_type_name"`
	Statuses   []CIStatusDistribution `json:"statuses"`
	TotalCIs   int64                 `json:"total_cis"`
}

// Helper methods

// NewLifecycleStatusFromRequest creates a new LifecycleStatus from a CreateLifecycleStatusRequest
func NewLifecycleStatusFromRequest(req *CreateLifecycleStatusRequest, createdBy uuid.UUID) *LifecycleStatus {
	now := time.Now()
	sortOrder := 0
	if req.SortOrder != nil {
		sortOrder = *req.SortOrder
	}

	return &LifecycleStatus{
		ID:          uuid.New(),
		Name:        req.Name,
		DisplayName: req.DisplayName,
		Description: req.Description,
		Color:       req.Color,
		Icon:        req.Icon,
		SortOrder:   sortOrder,
		IsActive:    true,
		IsSystem:    false,
		CreatedBy:   createdBy,
		CreatedAt:   now,
	}
}

// UpdateFromRequest updates a LifecycleStatus from an UpdateLifecycleStatusRequest
func (ls *LifecycleStatus) UpdateFromRequest(req *UpdateLifecycleStatusRequest, updatedBy uuid.UUID) {
	if req.DisplayName != nil {
		ls.DisplayName = *req.DisplayName
	}
	if req.Description != nil {
		ls.Description = req.Description
	}
	if req.Color != nil {
		ls.Color = req.Color
	}
	if req.Icon != nil {
		ls.Icon = req.Icon
	}
	if req.SortOrder != nil {
		ls.SortOrder = *req.SortOrder
	}
	if req.IsActive != nil {
		ls.IsActive = *req.IsActive
	}

	ls.UpdatedBy = &updatedBy
	now := time.Now()
	ls.UpdatedAt = &now
}

// CanBeModified checks if the lifecycle status can be modified
func (ls *LifecycleStatus) CanBeModified() bool {
	return !ls.IsSystem
}

// CanBeDeleted checks if the lifecycle status can be deleted
func (ls *LifecycleStatus) CanBeDeleted() bool {
	return !ls.IsSystem
}

// GetDisplayColor returns the display color or a default
func (ls *LifecycleStatus) GetDisplayColor() string {
	if ls.Color != nil {
		return *ls.Color
	}
	return "#6b7280" // Default gray color
}

// GetDisplayIcon returns the display icon or a default
func (ls *LifecycleStatus) GetDisplayIcon() string {
	if ls.Icon != nil {
		return *ls.Icon
	}
	return "circle" // Default icon
}

// Validate checks if the lifecycle status is valid
func (ls *LifecycleStatus) Validate() error {
	if ls.Name == "" {
		return fmt.Errorf("name is required")
	}
	if ls.DisplayName == "" {
		return fmt.Errorf("display name is required")
	}
	if ls.Color != nil && *ls.Color != "" {
		if len(*ls.Color) != 7 || (*ls.Color)[0] != '#' {
			return fmt.Errorf("color must be a valid hex color code (#RRGGBB)")
		}
	}
	return nil
}